### PR TITLE
fix: CNS-948: fixed mix pairing bug

### PR DESCRIPTION
**Pull request by oren-lava**
This PR fixes a bug of pairing with MIXED selected providers mode. The bug can be explained with the following scenario:
 
The strictest policy indicates that the selected providers should have both "debug" extension and "archive" addon
There is only one provider that support both "debug" and "archive". There are two more providers: one supports only "debug" and the other only supports "archive".

The mix filters code had a bug that if the policy requests "debug" + "archive" and there were only providers that support either "debug" or "archive", they wouldn't necessarily be picked over "normal" providers that don't support either.

Code behavior with bug: pairing includes the provider that supports both "debug" + "archive" and 3 random providers.
Code behavior without bug: pairing includes the provider that supports both "debug" + "archive", the "debug" provider, the "archive" provider and one random provider.

### CODE REVIEW

File: x/pairing/keeper/filters/addon_filter.go [hunk 1]
```diff
@@ -26,22 +29,73 @@ func (f *AddonFilter) InitFilter(strictestPolicy planstypes.Policy) bool {
...
+				}
+				return true, filters
 			}
 		}
 	}
-	return false
+	return false, nil
+}
+
+func convertAddonFilterToFilter(af AddonFilter) Filter {
```
Comment (Yaroms) [inline x/pairing/keeper/filters/addon_filter.go:97]: why do we need a method for this?

File: x/pairing/keeper/filters/filter.go [hunk 2]
```diff
@@ -27,13 +44,16 @@ func GetAllFilters() []Filter {
...
-func initFilters(filters []Filter, strictestPolicy planstypes.Policy) []Filter {
+func initFilters(filters []Filter, strictestPolicy planstypes.Policy) (originalFilters []Filter) {
 	activeFilters := []Filter{}
 
 	for _, filter := range filters {
-		active := filter.InitFilter(strictestPolicy)
+		active, subMixFilters := filter.InitFilter(strictestPolicy)
 		if active {
 			activeFilters = append(activeFilters, filter)
+			if subMixFilters != nil {
```
Comment (Yaroms) [inline x/pairing/keeper/filters/filter.go:54]: appending to a list while iterating, looks ambiguous

Comment (Yaroms) [reply x/pairing/keeper/filters/filter.go:54]: nvm

File: x/pairing/keeper/filters/filter.go [hunk 3]
```diff
@@ -27,13 +44,16 @@ func GetAllFilters() []Filter {
 	return filters
 }
 
-func initFilters(filters []Filter, strictestPolicy planstypes.Policy) []Filter {
+func initFilters(filters []Filter, strictestPolicy planstypes.Policy) (originalFilters []Filter) {
```
Comment (omerlavanet) [inline x/pairing/keeper/filters/filter.go:47]: they are not original though, also the mixed filters are there

File: x/pairing/keeper/pairing_test.go [hunk 4]
```diff
@@ -2039,6 +2014,134 @@ func TestExtensionAndAddonPairing(t *testing.T) {
...
+//   - There is only one provider that support both "ext1" extension and "addon" addon. There are two more providers
+//     one supports only "ext1" and the other only supports "addon".
+//
+//     The mix filters code had a bug that if the policy requests ext1+addon and there were only providers that support either
+//     "ext1" or "addon", they wouldn't necessarily be picked over "normal" providers.
+//
+//     Test result with bug: pairing includes the provider that supports both and 3 random providers
+//     Test result without bug: pairing includes the provider that supports both, the "ext1" provider, the "addon" provider
+//     and one random provider
+func TestMixBothExentionAndAddonPairingBug(t *testing.T) {
```
Comment (omerlavanet) [inline x/pairing/keeper/pairing_test.go:2030]: it wasn't a bug just a strict requirement, let's remove the wording of bug from here :)

**Review by omerlavanet — APPROVED** [review]

File: x/pairing/keeper/filters/filter.go (hunk 3 above)
Comment (oren-lava) [reply x/pairing/keeper/filters/filter.go]: done. see https://github.com/lavanet/lava/commit/7e3610f4d558155f11daf3e24fa73afa58b06eea

File: x/pairing/keeper/pairing_test.go (hunk 4 above)
Comment (oren-lava) [reply x/pairing/keeper/pairing_test.go]: done. see https://github.com/lavanet/lava/commit/7e3610f4d558155f11daf3e24fa73afa58b06eea

**Review by Yaroms — APPROVED** [review]

**Pull request merged**